selection error diagnostics results in a query cycle

when fetching `effective_visibilities` for `Adt` self types of a local trait while proving a where-clause.

https://github.com/rust-lang/rust/blob/8c32e313cccf7df531e2d49ffb8227bb92304aee/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs#L2871-L2889

This query fetches the hidden type of all opaque types https://github.com/rust-lang/rust/blob/8c32e313cccf7df531e2d49ffb8227bb92304aee/compiler/rustc_privacy/src/lib.rs#L1841-L1846

introduced in #143431, cc @xizheyin @compiler-errors

```rust
trait LocalTrait {}
struct SomeType;
fn impls_trait<T: LocalTrait>() {}
fn foo() -> impl Sized {
    impls_trait::<SomeType>();
}
```

results in
```
error[E0391]: cycle detected when computing type of `foo::{opaque#0}`
 --> src/lib.rs:4:13
  |
4 | fn foo() -> impl Sized {
  |             ^^^^^^^^^^
  |
note: ...which requires computing type of opaque `foo::{opaque#0}`...
 --> src/lib.rs:4:13
  |
4 | fn foo() -> impl Sized {
  |             ^^^^^^^^^^
note: ...which requires borrow-checking `foo`...
 --> src/lib.rs:4:1
  |
4 | fn foo() -> impl Sized {
  | ^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `foo`...
 --> src/lib.rs:4:1
  |
4 | fn foo() -> impl Sized {
  | ^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires checking if `foo` contains FFI-unwind calls...
 --> src/lib.rs:4:1
  |
4 | fn foo() -> impl Sized {
  | ^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires building MIR for `foo`...
 --> src/lib.rs:4:1
  |
4 | fn foo() -> impl Sized {
  | ^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires match-checking `foo`...
 --> src/lib.rs:4:1
  |
4 | fn foo() -> impl Sized {
  | ^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires type-checking `foo`...
 --> src/lib.rs:4:1
  |
4 | fn foo() -> impl Sized {
  | ^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which requires checking effective visibilities...
  = note: ...which again requires computing type of `foo::{opaque#0}`, completing the cycle
note: cycle used when checking that `foo::{opaque#0}` is well-formed
 --> src/lib.rs:4:13
  |
4 | fn foo() -> impl Sized {
  |             ^^^^^^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

error[E0277]: the trait bound `SomeType: LocalTrait` is not satisfied
 --> src/lib.rs:5:19
  |
5 |     impls_trait::<SomeType>();
  |                   ^^^^^^^^ the trait `LocalTrait` is not implemented for `SomeType`
  |
help: this trait has no implementations, consider adding one
 --> src/lib.rs:1:1
  |
1 | trait LocalTrait {}
  | ^^^^^^^^^^^^^^^^
note: required by a bound in `impls_trait`
 --> src/lib.rs:3:19
  |
3 | fn impls_trait<T: LocalTrait>() {}
  |                   ^^^^^^^^^^ required by this bound in `impls_trait`
```
note the query cycle before the actually desired error

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -2878,7 +2878,7 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                                 // we check if `TraitB` can be reachable from `S`
                                 // to determine whether to note `TraitA` is sealed trait.
                                 if let ty::Adt(adt, _) = ty.kind() {
-                                    let visibilities = tcx.effective_visibilities(());
+                                    let visibilities = &tcx.resolutions(()).effective_visibilities;
                                     visibilities.effective_vis(local).is_none_or(|v| {
                                         v.at_level(Level::Reexported)
                                             .is_accessible_from(adt.did(), tcx)

--- a/tests/ui/trait-bounds/trait-bound-adt-issue-145611.rs
+++ b/tests/ui/trait-bounds/trait-bound-adt-issue-145611.rs
@@ -1,0 +1,11 @@
+// This test is for regression of issue #145611
+// There should not be cycle error in effective_visibilities query.
+
+trait LocalTrait {}
+struct SomeType;
+fn impls_trait<T: LocalTrait>() {}
+fn foo() -> impl Sized {
+    impls_trait::<SomeType>(); //~ ERROR the trait bound `SomeType: LocalTrait` is not satisfied [E0277]
+}
+
+fn main() {}

--- a/tests/ui/trait-bounds/trait-bound-adt-issue-145611.stderr
+++ b/tests/ui/trait-bounds/trait-bound-adt-issue-145611.stderr
@@ -1,0 +1,20 @@
+error[E0277]: the trait bound `SomeType: LocalTrait` is not satisfied
+  --> $DIR/trait-bound-adt-issue-145611.rs:8:19
+   |
+LL |     impls_trait::<SomeType>();
+   |                   ^^^^^^^^ the trait `LocalTrait` is not implemented for `SomeType`
+   |
+help: this trait has no implementations, consider adding one
+  --> $DIR/trait-bound-adt-issue-145611.rs:4:1
+   |
+LL | trait LocalTrait {}
+   | ^^^^^^^^^^^^^^^^
+note: required by a bound in `impls_trait`
+  --> $DIR/trait-bound-adt-issue-145611.rs:6:19
+   |
+LL | fn impls_trait<T: LocalTrait>() {}
+   |                   ^^^^^^^^^^ required by this bound in `impls_trait`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

